keps/sig-cluster-lifecycle/kubeadm/20190506-kubeadm-machine-output.md

```diff
@@ -0,0 +1,363 @@
+
+## Release Signoff Checklist
+
+**ACTION REQUIRED:** In order to merge code into a release, there must be an issue in [kubernetes/enhancements] referencing this KEP and targeting a release milestone **before [Enhancement Freeze](https://github.com/kubernetes/sig-release/tree/master/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core Kubernetes i.e., [kubernetes/kubernetes], we require the following Release Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These checklist items _must_ be updated for the enhancement to be released.
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Kubeadm should support structured output, such as JSON, YAML, or a [Go template](https://golang.org/pkg/text/template/).
+
+## Motivation
+
+While Kubernetes may be deployed manually, the de facto, if not de jeur, means of turning up a Kubernetes cluster is kubeadm. Popular systems management software, such as Terraform, rely on kubeadm in order to deploy Kubernetes. Planned enhancements to the Cluster API project include a composable package for bootstrapping Kubernetes with kubeadm and cloud-init.
+
+Without structured output, even the most seemingly innocuous changes could break Terraform, Cluster API, or other software that relies on the results of kubeadm.
+
+### Goals
+
+* Kubeadm should support structured output, including, but not limited to, the following output types:
+  * Buffered
+    * JSON
+    * YAML
```

Comment:
I understand YAML output for commands which output data designed to be modified by users, but I don't feel like YAML is going to be a useful format for this style of output.

Users of machine readable output typically want the strictest and fastest output formats possible, which out of these options is JSON.

Other than this, +1 for this overall KEP as far as minikube is concerned.

Reply:
Hi @tstromberg,

> If YAML support is necessary, you may want to describe the rationale more specifically than "kubectl does it".

Do we though? The proposed design gives YAML basically for free since it's just another means of encoding the data. And if we use `cli-runtime` then YAML is already included. At that point we'd be removing the option that's already there...

Reply:
Hi @tstromberg,

Any thoughts here?

Reply:
If YAML comes for free/with low effort, I'm ok in supporting it because it is quite common in K8s world (eeetoooo much yaml!)